File: Week1_Part2/Vigenere/vigenere.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>

int main(int argc, char* argv[])
{
    if (argc != 2)
    {
        printf("Please give a key to encrypt with!\n");
        return 1;
    }
    for (int i = 0; i < strlen(argv[1]); i++)
    {
        if (!isalpha(argv[1][i]))
        {
            printf("Your key should only consist of letters!");
            return 1;
        }
    }
    char* key = argv[1];
    char* plaintext = get_string("Give me a string to cipher: ");
    char* ciphertext = (char*)malloc((strlen(plaintext) + 1) * sizeof(char));
    
    int keyLength = strlen(key);

    for (int i = 0, k = 0; i < strlen(plaintext); i++)
    {
        if (isalpha(plaintext[i]))
        {
            int keychar = (toupper(key[k++ % keyLength]) - 'A') % 26;

            if (isupper(plaintext[i]))
            {
                ciphertext[i] = ((plaintext[i] + keychar) - 'A') % 26 + 'A';
            }
            else
            {  
                ciphertext[i] = ((plaintext[i] + keychar) - 'a') % 26 + 'a';
            }
        }
        else
        {
            ciphertext[i] = plaintext[i];
        }
    }
    ciphertext[strlen(plaintext)] = '\0';
    printf("%s: " "%s\n", "plaintext", plaintext);
    printf("%s: " "%s\n", "ciphertext", ciphertext);

    free(ciphertext);
}

File: Week5/Sentimental/vigenere/vigenere.py
import sys


ALPHABET_LENGTH = 26
ASCII_UPPERCASE_A = 65
ASCII_LOWERCASE_A = 97


def encryptWord(plainWord, key):
    encryptedWord = []
    keyLength = len(key)
    keyIndex = 0

    for i in range(len(plainWord)):
        if plainWord[i].isalpha():
            keycharInt = ord((key[keyIndex % keyLength].upper()))
            keychar = (keycharInt - ASCII_UPPERCASE_A) % 26

            if plainWord[i].isupper():
                encryptedLetter = getEncryptedLetter(ASCII_UPPERCASE_A, keychar, plainWord[i])
                encryptedWord.append(encryptedLetter)
            else:
                encryptedLetter = getEncryptedLetter(ASCII_LOWERCASE_A, keychar, plainWord[i])
                encryptedWord.append(encryptedLetter)
            keyIndex += 1
        else:
            encryptedWord.append(plainWord[i])

    return ''.join(encryptedWord)


def getEncryptedLetter(asciiCaseA, keychar, plainWordChar):
    letter = str(chr(((ord(plainWordChar) + keychar) - asciiCaseA) % ALPHABET_LENGTH + asciiCaseA))
    return letter


def main():
    if len(sys.argv) != 2:
        print("Give me one key to encript with! ")
        exit(1)

    key = sys.argv[1]
    if not key.isalpha():
        print("Your key should only consist of letters!")
        exit(1)

    plaintext = input("Give me a word to cipher!: ")
    encryptedWord = encryptWord(plaintext, key)

    print("plaintext: ", plaintext)
    print("ciphertext: ", encryptedWord)

if __name__ == '__main__':
    main()
